---
title: "Lecture 11: Comparison and prediction of protein structure"
author:
- affiliation: Fundación ARAID & Estación Experimental de Aula Dei-CSIC, Zaragoza
    España <http://www.eead.csic.es/compbio>
  name: Bruno Contreras Moreira
date: '`r format(Sys.Date(),"%d/%b/%Y")`'
output:
  html_document:
    fig_caption: yes
    highlight: zenburn
    number_sections: yes
    theme: cerulean
    toc: yes
    toc_float: yes
  ioslides_presentation:
    css: ./comm/bkgr.css
    fig_caption: yes
    incremental: yes
    logo: ./comm/quantBiotech.png
    smaller: yes
    toc: yes
    toc_depth: 3
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  slidy_presentation:
    incremental: yes
  word_document: default
  html_notebook: 
    toc: yes
geometry: margin=1cm
fontsize: 11pt
bibliography: bib_myarticles_md.bib
---

```{r knitr setup, include=FALSE,  eval=TRUE, echo=FALSE, warning=FALSE}
library(knitr)
knitr::opts_chunk$set(eval=TRUE, cache=FALSE, message=FALSE, warning=FALSE, 
                      comment = "", results="markup")
```

# Introduction

Previous lectures focused on genomes and some algorithms used to study them. Often the most valuable product of genomic projects are the annotations of genes, particularly those that encode proteins that are actually expressed in relevant tissues or conditions. Even if their sequences are known, their function might not be obvious, as sequences can diverge in evolution so that homologies are difficult to track. However, we have learned that the structure of proteins is more conserved than their sequence. Therefore, by using structure similarity it is frequently possible to annotate sequences with few similar sequences in the database. The goal of this lecture is to describe current approaches for the study of protein structure, and how they can be used to annotate protein function.

# Protein structure 

Protein structure has been classically handled as four suporimposed layers, starting from the primary structure (the secuence) up to the quaternary structure, when several polypeptide chains interact. 

## Primary structure 

Primary structure is the linear sequence of amino acid residues encoded in an open reading frame.
It is usually represented as a chain where each letter identifies a residue. For instance, the following FASTA snippet describes 30 amino acid residues of fruit fly insuline:

```{r primary, eval=FALSE}
>Insuline [Drosophila melanogaster]
MFSQHNGAAV HGLRLQSLLI AAMLTAAMAM...
```

In this one-letter code M corresponds to methionine, Q to glutamine and so on, just as in BLOSUM matrices:

![**Table.** Summary of natural amino acid residues normally found in proteins, taken from [http://www.ionsource.com/virtit/VirtualIT/aainfo.htm](http://www.ionsource.com/virtit/VirtualIT/aainfo.htm)](./pics/AAtable.png)

![**Figure.** Overlapping classifications of amino acids, taken from
[http://www.russelllab.org/aas](http://www.russelllab.org/aas)](./pics/AAVenn.png)

<!-- ![](./pics/blosum50.png) -->

## Secondary structure

Proteins in solution adopt conformations that maximize hydrogen bonds to neutralize the polar charges of the peptide backbone. In order to achieve this, two main types of secondary structure appear, right-hand $\alpha$-helices and $\beta$-sheets, connected by different kinds of turns.

![**Figure.** Example $\alpha$-helix, taken from [http://www.chembio.uoguelph.ca](http://www.chembio.uoguelph.ca)](./pics/est2protH.png)


![**Figure.** Example $\alpha$-helix with side-chains as sphere models and hydrogen bonds plotted as blue lines, taken from [http://structuralbioinformatics.com](http://structuralbioinformatics.com)](./pics/est2protH2.jpg)


![**Figure.** Possible conformations of $\beta$-sheets, taken from [http://www.chembio.uoguelph.ca](http://www.chembio.uoguelph.ca)](./pics/est2protE.jpg)


![**Figure.** Anti-parallel $\beta$-sheets with side-chains pointing out, taken from [http://structuralbioinformatics.com](http://structuralbioinformatics.com)](./pics/est2protE2.jpg)


Protein secondary structure (SS) can be encoded in a string just as the primary sequence (see the [DSSP code](https://en.wikipedia.org/wiki/Protein_secondary_structure#DSSP_classification)), assigning each residue a single letter associated to a particular SS state. Usually three states are used: $H$, used to call $\alpha$-helix residues, $E$ for extended sheet conformation and $C$ for the remaining *coil* residues. Following this convention the previous primary sequence could be annotated with the corresponding SS:

```{r 2ary, eval=FALSE}
MFSQHNGAAV HGLRLQSLLI AAMLTAAMAM...
EEEECCEEEE HHHHHHHHHH CCCCCCCCCC...
```

## Tertiary & quaternary structure

Most proteins fold like globules made of secondary structure elements connected by loops.
Folding (and evolutionary) units are usually called **domains** [@AnatomyProt2007, @Porter2012].
Domains tend to bury hydrophobic side-chains, exposing loops to the solvent [@Isom2010, @Branden1999]. Of course this is different for membrane or amyloid proteins [@Schnabel2010].

![**Figure.** Tertiary structure of enzyme thiosterase, with yellow $\beta$-sheets and pink $\alpha$-helices. Figure produced with [RasMol](http://rasmol.org)](./pics/1bvq.png)


![**Figure.** CIRCOS-like representation of tertiary structure, created with  [PDBCirclePlot](http://sacan.biomed.drexel.edu/pdbcircleplot)](./pics/1crn.png)

The coordinates and annotations of protein structures, often bound to other molecules, are available from a world-wide repository called the [Protein Data Bank](https://www.rcsb.org). 
R package [Rpdb](https://cran.r-project.org/web/packages/Rpdb) can be used to load and manipulate coordinate files in [PDB format](http://www.wwpdb.org/documentation/file-format):

```{r Rpdb, eval=FALSE}
library(R.utils) # for gzip
library(Rpdb) 

gunzip("test_data/test.pdb.gz")
pdb1 = read.pdb(file="test_data/test.pdb")
visualize(pdb1)

gzip("test_data/test.pdb")
```

Another R package [bio3D](https://cran.r-project.org/web/packages/bio3d) can be used to analyze PDB files, for instance to compute contact matrices:

```{r bio3d}
library(R.utils) # for gzip
library(bio3d) 

gunzip("test_data/test.pdb.gz")
pdb1 = read.pdb(file="test_data/test.pdb")

# make index of Calpha atoms
CAidx <- atom.select(pdb1, "calpha")

# contact cutoff is 6 angstrom
# residues <3 residues apart are skipped
contact_map = cmap( pdb1$xyz[CAidx$xyz], dcut=6, scut=3 )

plot.cmap(contact_map)

gzip("test_data/test.pdb")
```


The first place to explore interactively PDB structures is the [Protein Data Bank](https://www.rcsb.org) site. 

![**Figure.** 3D View of a structure in the Protein Data bank](./pics/3DView.png)


## Classifications

Classifications of protein structure are done out mainly at the domain level, 
as in [SCOP](http://scop.berkeley.edu) or [CATH](http://www.cathdb.info) databases,
although other schemes have been proposed, such as the periodic table of @Taylor2002, 
where the basic folding unit is a super-secondary structure supported by contacts among 
co-evolving residues [@Mackenzie2017,@Granata2017].

![**Figure.** Overview of the four main fold classes in SCOP, from @Hou2005](./pics/foldclassification.jpg)

![**Figure.** Periodic table of globular domains, from @Taylor2002](./pics/fold_periodic_table.jpg)


![**Figure.** Possible evolutionary paths connecting five folds in SCOP, from @Nepomnyachiy2014](./pics/domain_walk.jpg)

## Computing structural similarity

@Chothia1986 analyzed the relatioship between sequence conservation and structure divergence in proteins and found that structure is more conserved than sequence, as summarized in the next plot:

![**Figure.** Correlation between sequence and structure conservation in proteins, from @Chothia1986](./pics/chothia_lesk.jpg)

This seminal work reported that the sequence identity of two aligned protein sequences A and B can be used to estimate the structural divergence of their structures, computed in terms of root-mean square deviations ([RMSD](http://en.wikipedia.org/wiki/Root_mean_square_deviation), in Angstrom). 

$RMSD = \sqrt \frac{\sum_{i=1}^n dist(C\alpha_{i}^A - C\alpha_{i}^B)^2}{n}$

This and following work [see for instance @Illergard2009] suggest that protein structure is generally more conserved than sequence in Evolution. Furthermore, it has been seen that the structural effect of aminoacid substitution depends on how buried they are: mutations exposed to the solvent usually induce smaller structural divergences than those within the folding core.

<!-- ![**Figure.** Structural conservation of thioredoxins separated 4 billion years in Evolution, from @InglesPrieto2013](./pics/thioredoxins.jpg) -->

The most frequent ways to compute structural similarity require the computation of structural alignments, which are then used to compute similarity scores. This usually requires some kind of **dynamic programming** approach (see different algorithms [here](http://eead-csic-compbio.github.io/bioinformatica_estructural/node32.html), in Spanish), in which substitution matrices are replaced by structure-based scores that capture local similarity within a protein fold. The next diagram shows how this is done in the sotware SSAP [@Taylor1989], but there are many other approaches:

![**Figure.** Illustration of the atom-to-atom vectors calculated in SSAP to compute local, from  [https://en.wikipedia.org/wiki/Structural_alignment](https://en.wikipedia.org/wiki/Structural_alignment)](./pics/Ssapvectors.png)

Once the alignment and the corresponding superposition are computed, structural similarity scores can be computed such as RMSD, usally computed between $C\alpha$ of aligned residues, or E-values (see for instance software [MAMMOTH](https://ub.cbm.uam.es/software/mammoth.php)). Another popular score is TM-score, proposed by @Zhang2004, which is length-independent and is computed as follows:

$TMscore = max[  \frac{1}{L_{Q}}  \sum_{i=1}^{L_{T}} \frac{1}{ 1 + (\frac{d_{i}}{d_{0}})^{2} } ]$

Here $max$ is the maximum value after optimal superposition, $L_{Q}$ is the length of the query structure, $L_{T}$ is the number of residues aligned to the template structure, $d_{i}$ is the distance between the $i$th pair of residues and $d_{0}$ is a length-scaling factor. 
$TMscore$ takes values between 0 and 1. 
Benchmark experiments show that values of $TMscore > 0.5$ is typical of structures with the same fold [@Xu2010].

## Sequence-based classification: protein families

The annotation and comparison of protein structures, as well as the construction of profiles from their sequences and their homologues, are the cornerstones of classifications of protein families and domains. Both structural and sequence data are integrated in manually curated sequence profiles, which are then added protein sequences from available genomes and modelled as Hidden Markov Models (HMM). This is how [Pfam database](http://pfam.xfam.org) works.


![**Figure.** Domain composition of protein sequences from the Piwi family in Pfam](./pics/PfamDomains.png)


<!-- --------------------------------- -->

# Prediction algorithms

## Secondary structure can be predicted by analyzing natural protein variants

Sequence profiles obtained with no supervision with software such as PSI-BLAST or [HHblits](https://toolkit.tuebingen.mpg.de) have been used with success to predict secondary structure with help from Machine Learning tools, mainly neural networks. The available methods essentially learn how to recognize secondary structure elements from the patterns of amino acid substitutions. The first methods worked with 3 SS states (H,E,C, @Jones1999) but current methods are yielding reliable predictions in 8-states [@Jurtz2017].

![**Figure.** Three-state secondary structure can be predicted from PSI-BLAST profiles, from @Jones1999](./pics/psipredflow.gif)

## Protein folds can be recognized with sequence profiles

Sequence profiles of proteins for which structure data is available are successfully used to predict the fold of remote homologues, sequences with very low sequence similarity that often fail to be picked by standard tools such as BLASTP. For this to work a sequence profile of the query protein must be computed, for instance as a HMM, and then aligned to a non-redundant collection of HMMs of structures from the PDB. These alignments are aided by secondary structure predictions. This is how the [HHpred](https://toolkit.tuebingen.mpg.de) algorithm works [@Soding2005].

Protocols for fold recognition are often called [**threading algorithms**](http://en.wikipedia.org/wiki/Threading_\%28protein_sequence\%29), as they can be seen as protocols where the query sequence is threaded along a collection of folds. This is how

![**Figure.** Fold recognition algorithm, from @Guo2008](./pics/threading.jpg)

## Comparative modelling 

Once one or more templates have been selected, the structure of a query sequence can be modelled following the protocol of **comparative or homology modelling**. This is summarized in the next figure and relies on good quality alignments between both the query and the template sequence, which again are aided by secondary structure predictions. The alignments serve the purpose of copying coordinates of the peptide backbone from the template to the query. Any remaining gaps must be filled, but the core of the structure is inherited from the template.

![**Figure.** Comparative modelling protocol, from @Fiser2003](./pics/CMflow.png)






<!-- ## Prediction of mutant phenotypes

Mencionar SIFT, pero calcular con R contactos y mirar figuras de Baker -->

# References